geth/jail/jail_cell_test.go

```diff
@@ -109,3 +107,190 @@ func (s *JailTestSuite) TestJailLoopInCall() {
 		require.Fail("Failed to received event response")
 	}
 }
+
+func (s *JailTestSuite) TestJailLoopRace() {
+	require := s.Require()
+	require.NotNil(s.jail)
+
+	cell, err := s.jail.NewCell(testChatID)
+	require.NoError(err)
+	require.NotNil(cell)
+
+	items := make(chan string)
+
+	err = cell.Set("__captureResponse", func(val string) otto.Value {
+		go func() { items <- val }()
+		return otto.UndefinedValue()
+	})
+	require.NoError(err)
+
+	_, err = cell.Run(`
+		function callRunner(namespace){
+			return setTimeout(function(){
+				__captureResponse(namespace);
+			}, 1000);
+		}
+	`)
+	require.NoError(err)
+
+	for i := 0; i < 100; i++ {
+		_, err = cell.Call("callRunner", nil, "softball")
+		require.NoError(err)
+	}
+
+	for i := 0; i < 100; i++ {
+		select {
+		case received := <-items:
+			require.Equal(received, "softball")
+			break
+		case <-time.After(5 * time.Second):
+			require.Fail("test timed out")
+		}
+	}
+}
+
+func (s *JailTestSuite) TestJailFetchPromise() {
+	body := `{"key": "value"}`
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.Header().Add("Content-Type", "application/json")
+		w.Write([]byte(body))
+	}))
+	defer server.Close()
+
+	require := s.Require()
+	require.NotNil(s.jail)
+
+	cell, err := s.jail.NewCell(testChatID)
+	require.NoError(err)
+	require.NotNil(cell)
+
+	dataCh := make(chan otto.Value, 1)
+	errCh := make(chan otto.Value, 1)
+
+	err = cell.Set("__captureSuccess", func(res otto.Value) { dataCh <- res })
+	require.NoError(err)
+	err = cell.Set("__captureError", func(res otto.Value) { errCh <- res })
+	require.NoError(err)
+
+	// run JS code for fetching valid URL
+	_, err = cell.Run(`fetch('` + server.URL + `').then(function(r) {
+		return r.text()
+	}).then(function(data) {
+		__captureSuccess(data)
+	}).catch(function (e) {
+		__captureError(e)
+	})`)
+	require.NoError(err)
+
+	select {
+	case data := <-dataCh:
```

Comment:
I think the same here about the channel contents.

Reply:
But this tests the whole `fetch`/`promise` JS code. It makes sense to check if data is valid.

Reply:
I'm not sure. I want to decouple these tests but if you see it as an all-in-one test, it's fine, let's leave it. At least, most our tests are of such nature.